zilliqa/src/consensus.rs

```diff
@@ -526,6 +527,28 @@ impl Consensus {
             return Ok(None);
         }
 
+        let now = SystemTime::now();
+        if !during_sync
+            && block.timestamp() + self.config.consensus.maximum_delay_for_received_block < now
+        {
+            let diff = now.duration_since(SystemTime::UNIX_EPOCH)?
+                - (block.timestamp().duration_since(SystemTime::UNIX_EPOCH)?
+                    + self.config.consensus.maximum_delay_for_received_block);
+            warn!(
+                "Rejecting block - It's too old, diff is {}ms",
+                diff.as_millis()
+            );
+            return Ok(None);
+        }
+
+        if block.gas_used() > self.config.consensus.eth_block_gas_limit {
```

Comment:
I don't like this test at all, I'm afraid.
 - It means that bar some odd situations (syncing from a preprepared archive), you can never reduce the block gas limit, or you will choke syncing old blocks which had gas_used() higher than your current block_gas_limit.
 - We won't be processing this block anyway if it didn't get a 2/3 stake vote. What is the point of having a number which, if it is below the number everyone else has, simply causes you to halt your node? You might as well simply read eth_block_gas_limit from the blocks being voted on by the committee.
 - If you ever pick a number which is higher than other folks' eth_block_gas_limit, you will produce proposals that others will reject, causing potentially spurious view changes and network problems.

If we mean to set a block gas limit for the network, we should put it in a contract, and simply never vote for blocks that have a block gas limit higher than the one we set. The only case in which this test makes sense, AFAICT, is that we avoid checking the votes for blocks with gas_used() > self.config.consensus.eth_block_gas_limit, and that will only save us if there is an attacker who is dumb enough to lie about everything else, but not block.gas_used()...

Reply:
The block header should include both `gas_used` and `gas_limit` (the latter is missing), and we should reject the proposal if `block.gas_used > block.gas_limit` and also if `block.gas_limit > self.config.consensus.eth_block_gas_limit`.

@rrw-zilliqa the `propose()` function is used by validators receiving new block proposals, it's not executed when nodes are syncing historical blocks, and validators shall check if the proposed block is valid according to the current network settings before they vote for it